src/main/java/com/synopsys/integration/blackduck/codelocation/signaturescanner/command/ScanCommand.java

```diff
@@ -87,188 +94,225 @@ public ScanCommand(File signatureScannerInstallDirectory, File outputDirectory,
-            populateOfflineProperties(logger, specificRunOutputDirectoryPath, cmd);
+            populateOfflineProperties(logger, specificRunOutputDirectoryPath);
         }
 
         if (verbose) {
-            cmd.add("-v");
+            appendSingleArgument("-v");
         }
         if (debug) {
-            cmd.add("--debug");
+            appendSingleArgument("--debug");
         }
 
-        cmd.add("--logDir");
-        cmd.add(specificRunOutputDirectoryPath);
+        appendKeyValuePair("--logDir", specificRunOutputDirectoryPath);
 
         // Only add the statusWriteDir option if Black Duck supports the statusWriteDir option
         // The scanStatusDirectoryPath is the same as the log directory path
         // The CLI will create a subdirectory for the status files
-        cmd.add("--statusWriteDir");
-        cmd.add(specificRunOutputDirectoryPath);
+        appendKeyValuePair("--statusWriteDir", specificRunOutputDirectoryPath);
 
-        populateProjectAndVersion(cmd);
+        populateProjectAndVersion();
 
         if (StringUtils.isNotBlank(name)) {
-            cmd.add("--name");
-            cmd.add(name);
+            appendKeyValuePair("--name", name);
         }
 
-        populateExcludePatterns(cmd);
+        populateExcludePatterns();
 
         if (null != individualFileMatching) {
-            cmd.add("--individualFileMatching=" + individualFileMatching);
+            appendKeyValuePair("--individualFileMatching", individualFileMatching.toString());
         }
 
         if (isRapid) {
-            cmd.add("--no-persistence");
+            appendSingleArgument("--no-persistence");
             
             // --no-persistence-mode should never be used without --no-persistence so
             // only set it in this block.
-            cmd.add("--no-persistence-mode=" + bomCompareMode);
+            appendKeyValuePair("--no-persistence-mode", bomCompareMode);
         }
 
-        populateReducedPersistence(cmd);
+        populateReducedPersistence();
 
         if (StringUtils.isNotBlank(correlationId)) {
-            cmd.add("--correlationId");
-            cmd.add(correlationId);
+            appendKeyValuePair("--correlationId", correlationId);
         }
 
         ScanCommandArgumentParser parser = new ScanCommandArgumentParser();
-        populateAdditionalScanArguments(cmd, parser);
+        populateAdditionalScanArgumentsAndRemoveOverridden(parser);
+
+        return command;
+    }
+
+    private void appendSingleArgument(String arg) {
+        command.add(arg);
+    }
+
+    private void appendKeyValuePair(String key, String value) {
+        if (!commandKeysToKeyRelatedIndices.containsKey(key)) {
+            commandKeysToKeyRelatedIndices.put(key, new ArrayList<>());
+        }
+        List<Integer> indecesList = commandKeysToKeyRelatedIndices.get(key);
 
-        return cmd;
+        indecesList.add(command.size());
+        indecesList.add(command.size() + 1);
+
+        appendSingleArgument(key);
```

Comment:
Should this be instead appendKeyValuePair(...)?

Reply:
Here we're already in the `appendKeyValuePair()` method.
Or are you suggesting we have another one with a different set of arguments?

Reply:
I meant should this method or perhaps an additional method take a stream of key-value pairs? (...)

Reply:
I see. I am not convinced that it would make the code clearer.
When I was putting this together I was actually torn between doing what I did (calling `appendSingleArgument()`) or adding items to the list directly at these lines.
Do you think the latter would be better?

Reply:
It is too early for this work to justify that level of complexity. We can revisit when we implement centralized configuration management.